test/int/framework/frameworkTestSupport.ts

```diff
@@ -0,0 +1,69 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import * as path from 'path';
+import { ExtendedDebugClient } from 'vscode-chrome-debug-core-testsupport';
+
+/*
+ * A collection of supporting classes/functions for running framework tests
+ */
+
+export interface ProjectSpecProps {
+    /** The root directory of the test project */
+    projectRoot: string;
+    /** Source files directory of the test project */
+    projectSrc?: string;
+    /** The directory from which to host the project for a test */
+    webRoot?: string;
+    /** The outfiles directory for the test project */
+    outFiles?: string;
+    /** The default launch configuration for the test project */
+    launchConfig?: any;
+    /** Port to use for the server */
+    port?: number;
+    /** Url to use for the project */
+    url?: string;
+}
+
+/**
+ * Specifies an integration test project (i.e. a project that will be launched and
+ * attached to in order to test the debug adapter)
+ */
+export class TestProjectSpec {
+
+    _props: ProjectSpecProps;
```

Comment:
My goal here was to allow some of the properties to be undefined, and then fill them in with sensible defaults, such that you only need to pass props that change from the default (e.g. if 9/10 test projects have the same setup, we don't have to bother passing everything every time, but then still have the freedom to change 1 or 2 things for that 1/10 case).

Definitely open to other suggestions as well.

Reply:
I'm just suggesting changing the types to reflect that's what you are doing already.
The constructor can have undefined in the parameters, so it uses constructor(props: ProjectSpecProps) which has optional values

in the instance variable prop all the values are defined, because you default to the default values if the user didn't supply it, so all the values will be defined, so the most accurate type would be MakePropertyRequired<ProjectSpecProps>

Reply:
Got it, I see what you're saying.